README.md
# grid-neighbors-1d

Get the 8 closest neighbors of a grid with edge wrapping from a 1d array

## Why?
![grid-neighbors](https://raw.githubusercontent.com/phugh/grid-neighbors-1d/master/grid-neighbors.png)

## Usage
Get the neighbors of cell 12 in a 5x5 grid:
```javascript
import { getNeighbors } from "grid-neighbors-1d";
const neighbors = getNeighbors(12, 5, 5);
console.log(neighbors); // [7, 8, 13, 18, 17, 16, 11, 6] - clockwise from north
```

`getNeighbors` returns an array of indexes of the cell immediately to the north of the chosen cell, then clockwise around the remaining cells:

* neighbors[0] = north neighbor
* neighbors[1] = north-east neighbor
* neighbors[2] = east neighbor
* neighbors[3] = south-east neighbor
* neighbors[4] = south neighbor
* neighbors[5] = south-west neighbor
* neighbors[6] = west neighbor
* neighbors[7] = north-west neighbor

This means you can use destructuring if your environment supports it:

```javascript
const [north, northEast, east, southEast, south, southWest, west, northWest] = getNeighbors(4, 10, 6);
```

A `Direction` object is also available for convenience:

```javascript
import { Direction } from "grid-neighbors-1d";
const neighbors = getNeighbors(12, 5, 5);
console.log(neighbors[Direction.NORTH]); // 7
```

If memory usage is not a concern and you are likely to call `getNeighbors` many times, you may want to precalculate the neighbors for a given grid size:

```javascript
import { generateNeighborLookup } from "grid-neighbors-1d";
const getNeighbors = generateNeighborLookup(5, 5);
const neighbors = getNeighbors(12);
console.log(neighbors); // [7, 8, 13, 18, 17, 16, 11, 6]
```

For your convenience, TypeScript type declarations (index.d.ts), a declaration map (index.d.ts.map), and a sourcemap (index.js.map) are included.

## License
&copy; 2019-23 [P. Hughes](https://www.phugh.es). All rights reserved.

Shared under the [MIT](https://choosealicense.com/licenses/mit/) license.


CHANGELOG.md
# Changelog

## 2.1.0
- Added `generateNeighborLookup` to precalculate neighbors when the function will be called many times with the same grid dimensions
- Improved edge case handling

## 2.0.0
- Convert code to a more modern ES6 module style, rewritten in Typescript
- Added Typescript features (tsconfig.json, index.d.mts, index.d.mts.map)
- Improved error handling
- Improved documentation
- Improved developer experience (sourcemap, type declaration, declaration map)
- Lots of Github and VSCode developer features added


test/index.test.mjs
import { getNeighbors } from "../dist/index.js";

const arrayHasSameValues = (actual, expected) => actual.filter((v, i) => expected[i] === v).length === expected.length;

const aRes = [7, 8, 13, 18, 17, 16, 11, 6];
const aTest = getNeighbors(12, 5, 5);
if (!arrayHasSameValues(aTest, aRes)) throw new Error("Test 1 failed");

const bRes = [9, 5, 10, 15, 19, 18, 13, 8];
const bTest = getNeighbors(14, 5, 5);
if (!arrayHasSameValues(bTest, bRes)) throw new Error("Test 2 failed");

const cRes = [24, 20, 0, 5, 9, 8, 3, 23];
const cTest = getNeighbors(4, 5, 5);
if (!arrayHasSameValues(cTest, cRes)) throw new Error("Test 3 failed");

try {
  const x = getNeighbors();
  if (x) throw new Error("Test 4 failed");
} catch (err) {
  if (!(err instanceof SyntaxError)) throw err;
}

console.log("All tests passed");


index.js
/**
 * grid-neighbors-1d
 * v1.0.0
 *
 * Get the 8 closest neighbors of a cell in a 2d grid when flattened to a 1d array
 *
 * Help me make this better:
 * https://github.com/phugh/grid-neighbors-1d
 *
 * (C) 2019-20 P. Hughes
 * Licence : Creative Commons Attribution-NonCommercial-ShareAlike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 *
 * Usage example:
 * const gn = require("grid-neighbors-1d");
 * const neighbors = gn(12,5,6); // i.e. get the neighbors of cell 12 in a grid 5 high by 6 wide
 * console.log(neighbors); // [7, 8, 13, 18, 17, 16, 11, 6] - clockwise from north
 * // i.e. [north, north east, east, south east, south, south west, west, north west]
 *
 * See README.md for help.
 *
 * @param {number} cell current cell
 * @param {number} width grid width
 * @param {number} height grid height
 * @return {Array<number>} [north, north east, east, south east, south, south west, west, north west];
 */

// eslint-disable-next-line no-extra-semi
;(() => {
  'use strict';

  /**
   * @function getNeighbors
   * @param {number} cell current cell
   * @param {number} width grid width
   * @param {number} height grid height
   * @return {Array<number>} [north, north east, east, south east, south, south west, west, north west];
   */
  const getNeighbors = (cell, width, height) => {
    if (cell == null || width == null || height == null) {
      throw new Error('Invald input, expected cell, width and height.');
    } else if (isNaN(cell) || isNaN(width) || isNaN(height)) {
      throw new TypeError('Invald input, expected numbers only.');
    }

    const SIZE = width * height; // Total cells

    if (SIZE < 9) {
      throw new Error(`grid-neighbors: Minimum grid size is 9 cells. Provided grid is ${SIZE} cells.`);
    } else if (cell >= SIZE) {
      throw new Error(`grid-neighbors: Cell reference "${cell}" out of bounds. Maximum reference is ${SIZE - 1}.`);
    }
    // Setup
    const LC = Math.floor(cell / width) * width;  // left most cell
    const RC = (LC + width) - 1;                  // right most cell
    const SIZE_MINUS_WIDTH = SIZE - width;
    const CELL_MINUS_WIDTH = cell - width;
    const CELL_PLUS_WIDTH = cell + width;
    const CELL_MOD_WIDTH = cell % width;
    const TOP_RIGHT = width - 1;
    const BOTTOM_RIGHT = SIZE - 1;

    // Directions
    let north;
    let south;
    let east;
    let west;
    let northWest;
    let northEast;
    let southEast;
    let southWest;

    // North
    if (CELL_MINUS_WIDTH < 0) {                   // TOP EDGE
      north = SIZE_MINUS_WIDTH + cell;
    } else {
      north = CELL_MINUS_WIDTH;
    }

    // South
    if (CELL_PLUS_WIDTH >= SIZE) {                // BOTTOM EDGE
      south = cell - LC;
    } else {
      south = CELL_PLUS_WIDTH;
    }

    // East, North-East, South-East
    if (CELL_MOD_WIDTH === TOP_RIGHT) {           // RIGHT EDGE
      east = LC;
      if (cell === TOP_RIGHT) {                   // top right corner
        northEast = SIZE_MINUS_WIDTH;
        southEast = LC + width;
      } else if (cell === BOTTOM_RIGHT) {         // bottom right corner
        northEast = LC - width;
        southEast = 0;
      } else {
        northEast = LC - width;
        southEast = LC + width;
      }
    } else {
      east = cell + 1;
      northEast = north + 1;
      southEast = south + 1;
    }

    // West, North-West, South-West
    if (CELL_MOD_WIDTH === 0) {                   // LEFT EDGE
      west = RC;
      if (cell === SIZE_MINUS_WIDTH) {            // bottom left corner
        northWest = cell - 1;
        southWest = TOP_RIGHT;
      } else if (cell === 0) {                    // top left corner
        northWest = BOTTOM_RIGHT;
        southWest = RC + width;
      } else {
        northWest = cell - 1;
        southWest = RC + width;
      }
    } else {
      west = cell - 1;
      northWest = north - 1;
      southWest = south - 1;
    }

    return [north, northEast, east, southEast, south, southWest, west, northWest];
  };

  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = getNeighbors;
    }
    exports.getNeighbors = getNeighbors;
    exports.getNeighbours = getNeighbors;
  }
})();


src/index.ts
/**
 * grid-neighbors-1d
 * v2.1.0
 *
 * Get the 8 closest neighbors of a cell in a 2d grid when flattened to a 1d array
 *
 * Help me make this better:
 * https://github.com/phugh/grid-neighbors-1d
 *
 * @module grid-neighbors-1d
 * @copyright 2019-23 P. Hughes. All rights reserved.
 * @license MIT
 */

export type Neighbors = [
  north: number,
  northEast: number,
  east: number,
  southEast: number,
  south: number,
  southWest: number,
  west: number,
  northWest: number,
];

export const Direction = {
  NORTH: 0,
  NORTH_EAST: 1,
  EAST: 2,
  SOUTH_EAST: 3,
  SOUTH: 4,
  SOUTH_WEST: 5,
  WEST: 6,
  NORTH_WEST: 7,
} as const;

/**
 * @function getNeighbors
 * @description Get the 8 closest neighbors of a cell in a 2d grid when flattened to a 1d array
 * @param cell the cell to get neighbors for
 * @param width grid width
 * @param height grid height
 * @return an array of cell references [north, north-east, east, south-east, south, south-west, west, north-west]
 * @throws {SyntaxError} if function arguments are undefined or not numbers
 * @throws {RangeError} if total grid size is less than 9 cells or cell reference is out of bounds
 * @example
 *  import { getNeighbors } from 'grid-neighbors-1d';
 *  const neighbors = getNeighbors(12,5,6); // i.e. get the neighbors of cell 12 in a grid 5 high by 6 wide
 *  console.log(neighbors); // [7, 8, 13, 18, 17, 16, 11, 6] - clockwise from north
 *  // i.e. [north, north-east, east, south-east, south, south-west, west, north-west]
 */
export const getNeighbors = (cell: number, width: number, height: number): Neighbors => {
  // Validate input
  if (cell == null || isNaN(cell) || cell < 0) {
    throw new SyntaxError("Expected cell to be a non-negative number.");
  }
  if (width == null || isNaN(width) || width < 0) {
    throw new SyntaxError("Expected width to be a non-negative number.");
  }
  if (height == null || isNaN(height) || height < 0) {
    throw new SyntaxError("Expected height to be a non-negative number.");
  }

  const CELL = Math.floor(cell);
  const WIDTH = Math.floor(width);
  const HEIGHT = Math.floor(height);

  const SIZE = WIDTH * HEIGHT;
  if (SIZE < 9) {
    throw new RangeError(`Minimum grid size is 9 cells. Provided grid (${WIDTH} X ${HEIGHT}) is ${SIZE} cells.`);
  } else if (CELL >= SIZE) {
    throw new RangeError(`Cell reference "${CELL}" out of bounds. Maximum is ${SIZE - 1}.`);
  }

  const ROW = Math.floor(CELL / WIDTH);
  const COLUMN = CELL % WIDTH;

  const LEFT_MOST_CELL = ROW * WIDTH;
  const RIGHT_MOST_CELL = LEFT_MOST_CELL + WIDTH - 1;

  const SIZE_MINUS_WIDTH = SIZE - WIDTH;
  const CELL_MINUS_WIDTH = CELL - WIDTH;
  const CELL_PLUS_WIDTH = CELL + WIDTH;

  const TOP_RIGHT_CELL_IDX = WIDTH - 1;
  const BOTTOM_RIGHT_CELL_IDX = SIZE - 1;

  const north = CELL_MINUS_WIDTH < 0 ? SIZE_MINUS_WIDTH + CELL : CELL_MINUS_WIDTH;
  const south = CELL_PLUS_WIDTH >= SIZE ? CELL - LEFT_MOST_CELL : CELL_PLUS_WIDTH;

  let northEast: number;
  let east: number;
  let southEast: number;
  let southWest: number;
  let west: number;
  let northWest: number;

  // East, North-East, South-East
  if (COLUMN === TOP_RIGHT_CELL_IDX) {
    // RIGHT EDGE
    east = LEFT_MOST_CELL;
    if (CELL === TOP_RIGHT_CELL_IDX) {
      // top right corner
      northEast = SIZE_MINUS_WIDTH;
      southEast = LEFT_MOST_CELL + WIDTH;
    } else if (CELL === BOTTOM_RIGHT_CELL_IDX) {
      // bottom right corner
      northEast = LEFT_MOST_CELL - WIDTH;
      southEast = 0;
    } else {
      northEast = LEFT_MOST_CELL - WIDTH;
      southEast = LEFT_MOST_CELL + WIDTH;
    }
  } else {
    east = CELL + 1;
    northEast = north + 1;
    southEast = south + 1;
  }

  // West, North-West, South-West
  if (COLUMN === 0) {
    // LEFT EDGE
    west = RIGHT_MOST_CELL;
    if (CELL === SIZE_MINUS_WIDTH) {
      // bottom left corner
      northWest = CELL - 1;
      southWest = TOP_RIGHT_CELL_IDX;
    } else if (CELL === 0) {
      // top left corner
      northWest = BOTTOM_RIGHT_CELL_IDX;
      southWest = RIGHT_MOST_CELL + WIDTH;
    } else {
      northWest = CELL - 1;
      southWest = RIGHT_MOST_CELL + WIDTH;
    }
  } else {
    west = CELL - 1;
    northWest = north - 1;
    southWest = south - 1;
  }

  return [north, northEast, east, southEast, south, southWest, west, northWest];
};

export const getNeighbours = getNeighbors;

/**
 * Pre-generate a lookup table for all cells in a grid
 * @param width grid width
 * @param height grid height
 * @returns a function that takes a cell reference and returns an array of cell references [north, north-east,....]
 * @throws {SyntaxError} if function arguments are undefined or not numbers
 * @throws {RangeError} if total grid size is less than 9 cells
 * @example
 * import { generateNeighborLookup } from 'grid-neighbors-1d';
 * const getNeighbors = generateNeighborLookup(5,6);
 * const neighbors = getNeighbors(12); // i.e. get the neighbors of cell 12 in a grid 5 high by 6 wide
 * console.log(neighbors); // [7, 8, 13, 18, 17, 16, 11, 6] - clockwise from north
 */
export const generateNeighborLookup = (width: number, height: number): ((cell: number) => Neighbors | undefined) => {
  const neighborLookup: { [key: number]: Neighbors } = {};
  const size = width * height;
  for (let cell = 0; cell < size; cell++) {
    neighborLookup[cell] = getNeighbors(cell, width, height);
  }
  return (cell: number) => neighborLookup[cell];
};

export const generateNeighbourLookup = generateNeighborLookup;
